src/v/datalake/translation/deps.cc

```diff
@@ -10,12 +10,519 @@
+
+    void close() noexcept final {
+        _partition->unregister_flush_hook(_partition_flush_subscription);
+    }
+
+    const model::ntp& ntp() const final { return _partition->ntp(); }
+
+    model::revision_id topic_revision() const final {
+        return _partition->get_topic_revision_id();
+    }
+
+    model::term_id term() const final { return _partition->term(); }
+
+    ss::future<kafka::offset> wait_for_data_to_translate(
+      std::optional<kafka::offset> last_translated_offset,
+      ss::abort_source& as) final {
+        // naively wait a few polling periods after some data appears, hopefully
+        // grabbing a few more batches before breaking back out to the
+        // translator
+        // TODO: probably choose a better value. configurable? or can we do it
+        // batchwise?
+        constexpr kafka::offset::type offsets_target = 10;
+        constexpr int iter_limit = 5;
+        kafka::offset::type n_to_translate = 0;
+        int n_poll = 0;
+        while (n_poll < iter_limit && n_to_translate < offsets_target) {
+            if (auto to_translate = has_more_data_to_translate(
+                  last_translated_offset);
+                to_translate.has_value()) {
+                ++n_poll;
+                n_to_translate = to_translate.value();
+            }
+            co_await ss::sleep_abortable(poll_duration, as);
+        }
+        auto read_begin_offset = min_offset_for_translation();
+        if (last_translated_offset) {
+            read_begin_offset = kafka::next_offset(
+              last_translated_offset.value());
+        }
+        co_return read_begin_offset;
+    }
+
+    ss::future<std::optional<model::record_batch_reader>> make_log_reader(
+      kafka::offset begin_offset,
+      ss::io_priority_class io_priority,
+      ss::abort_source& as) final {
+        auto max_translatable_offset = max_offset_for_translation();
+        if (
+          !max_translatable_offset
+          || max_translatable_offset.value() < begin_offset) {
+            co_return std::nullopt;
+        }
+        auto log_reader = co_await _partition_proxy->make_reader(
+          {kafka::offset_cast(begin_offset),
+           kafka::offset_cast(max_translatable_offset.value()),
+           0,
+           std::numeric_limits<size_t>::max(),
+           io_priority,
+           std::nullopt,
+           std::nullopt,
+           as});
+        auto tracker = kafka::aborted_transaction_tracker::create_default(
+          _partition_proxy.get(), std::move(log_reader.ot_state));
+        co_return model::make_record_batch_reader<kafka::read_committed_reader>(
+          std::move(tracker), std::move(log_reader.reader));
+    }
+
+    kafka::offset min_offset_for_translation() const final {
+        if (_partition->is_read_replica_mode_enabled()) {
+            return model::offset_cast(_partition_proxy->start_offset());
+        }
+        return model::offset_cast(_partition_proxy->local_start_offset());
```

Comment:
I think we should always use `_partition_proxy->start_offset()`. In case of node pool migration with fast partition movements we may introduce a gap as we may have to read from tiered storage

Reply:
I'll have to take your word for it, unfortunately. Will include in the next push, but @bharathv please chime in if you feel the need.

Reply:
Hm.. I'm curious how difficult it would be to address this in node pool migration? Changing this here would have pretty major repercussions for how we handle topic recovery and WCR (we've opted to not backup coordinator state, and rely on local-only translation to avoid re-translating the entire log)

Reply:
for posterity: discussed in group and concluded that it's not an issue in practice because we take this path iff we're starting up with no coordinator state. BUT it still might be a good idea to make fast partition movement contingent on max translatable offset.